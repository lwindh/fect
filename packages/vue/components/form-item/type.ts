import type { FormItemProps } from './props'

export const READONLY_FORM_ITEM_KEY = Symbol('FormItemKey')

export interface FormItemProvide {
  formItemProps: FormItemProps
  resetField: () => void
  clearValidate: () => void
}
