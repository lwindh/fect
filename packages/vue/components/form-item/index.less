.fect-form-item {
  display: flex;
  align-items: center;
  margin-bottom: calc(var(--fay-gap) * 1.25);

  &--inline {
    display: inline-flex;
    align-items: center;
    margin-right: var(--fay-gap-half);
  }
  &--top {
    display: block;
    // .fect-form-item__label {
    //   padding: 0 0 var(--fay-gap-half) 0;
    // }
    .fect-form-item__content {
      margin-top: var(--fay-gap-half);
    }
  }

  &__label {
    box-sizing: border-box;
    padding: 0 var(--fay-gap-half) 0 0;
    &.is-required {
      &::before {
        content: '*';
        color: var(--error-default);
        margin-right: 5px;
      }
    }
  }
  &__content {
    position: relative;
    flex: 1;
    min-width: 0;
  }
  &__error {
    color: var(--error-light);
    font-size: 13px;
    line-height: 1;
    padding-top: var(--fay-gap-quarter);
    position: absolute;
    top: 100%;
    left: 0;
    &--hidden {
      display: none;
    }
  }
}
