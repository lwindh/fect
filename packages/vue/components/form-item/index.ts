import _FormItem from './form-item'
import { withInstall } from '../utils'

export const FormItem = withInstall(_FormItem)

export default FormItem
