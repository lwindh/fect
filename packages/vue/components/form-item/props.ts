import { PropType, ExtractDefaultPropTypes } from 'vue'
import type { LabelPosition } from '../form/type'

export const props = {
  prop: String,
  for: {
    type: String,
    default: ''
  },
  label: String,
  labelWidth: {
    type: [String, Number],
    default: ''
  },
  labelPosition: {
    type: String as PropType<LabelPosition>,
    default: ''
  },
  required: Boolean,
  showMessage: Boolean
}

export type FormItemProps = ExtractDefaultPropTypes<typeof props>
