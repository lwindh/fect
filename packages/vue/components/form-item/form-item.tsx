import { defineComponent, ref, computed, reactive, toRefs } from 'vue'
import { useProvider, createProvider } from '@fect-ui/vue-hooks'
import { createName } from '../utils'
import { getLabelPostion, getLabelWidth, getFormItemLayoutClass, getUnVerfiedClass } from '../form/style'
import { READONLY_FORM_KEY, Trigger } from '../form/type'
import { READONLY_FORM_ITEM_KEY } from './type'
import { props } from './props'
import type { FormProvide } from '../form/type'
import type { CSSProperties } from 'vue'

import './index.less'

const name = createName('FormItem')

export default defineComponent({
  name,
  props,
  setup(props, { slots }) {
    const { context } = useProvider<FormProvide>(READONLY_FORM_KEY)

    const { provider } = createProvider(READONLY_FORM_ITEM_KEY)

    if (!context && process.env.NODE_ENV !== 'production') {
      console.error('[Fect] <FormItem /> must be a child component of <Form />')
      return
    }

    const formItemRef = ref<HTMLDivElement>()

    const setLabelStyle = computed(() => {
      const { formProps } = context!
      const style: CSSProperties = {}
      const labelPosition = getLabelPostion(props.labelPosition || formProps.labelPosition)
      if (!labelPosition) return style
      const labelWidth = getLabelWidth(props.labelWidth || formProps.labelWidth)
      style.width = labelWidth
      style.textAlign = labelPosition
      return style
    })

    const setClass = computed(() => {
      const { formProps } = context!
      const labelPosition = props.labelPosition || formProps.labelPosition
      const basisClass = 'fect-form-item'
      return getFormItemLayoutClass(formProps.inline, labelPosition, basisClass)
    })

    const formItemProps = reactive({ ...toRefs(props) })

    // const setUnVerfiedClass = computed(() => {
    //   const { formProps } = context!
    //   const display = props.showMessage || formProps.showMessage
    //   return getUnVerfiedClass(display, 'fect-form-item__error')
    // })

    /**
     * user may pass native attrs  for . so  we should intercept the attr
     */
    const setLabelFor = computed(() => props.prop || props.for)

    // ${setRequired.value ? 'is-required' : ''}
    return () => (
      <div class={`fect-form-item ${setClass.value}`} ref={formItemRef}>
        <label class={`fect-form-item__label `} for={setLabelFor.value} style={setLabelStyle.value}>
          {props.label}
        </label>
        <div class="fect-form-item__content">
          {slots.default?.()}
          {/* {false && (
            <div class={`fect-form-item__error ${setUnVerfiedClass.value}`}>
              Please choose the direction you are interested in
            </div>
          )} */}
        </div>
      </div>
    )
  }
})
