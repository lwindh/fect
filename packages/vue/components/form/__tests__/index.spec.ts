import { mount } from '@vue/test-utils'
import { Form } from '../index'
import { FormItem } from '../../form-item'

describe('Form', () => {
  it('props should work correctly', async () => {
    const wrapper = mount(Form, {
      props: {
        labelPosition: 'left',
        labelWidth: '110',
        inline: true
      },
      slots: {
        default: () => FormItem
      }
    })

    await wrapper.setProps({ inline: false })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should wrok normal', () => {
    const wrapper = mount({
      components: {
        [FormItem.name]: FormItem,
        [Form.name]: Form
      },
      template: `<fe-form label-width="120px">
         <fe-form-item label-width="100" label-position="top">1</fe-form-item>
         <fe-form-item label-width="10" label-position="left">2</fe-form-item>
         <fe-form-item >3</fe-form-item>
      </fe-form>`
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
  it('form-item only work in form ', () => {
    const wrapper = mount(FormItem, {})
    expect(() => wrapper.get('.fect-form-item')).toThrowError()
  })
})
