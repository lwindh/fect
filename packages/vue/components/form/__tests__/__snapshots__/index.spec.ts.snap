// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form props should work correctly 1`] = `<form class="fect-form "> </form>`;

exports[`Form should wrok normal 1`] = `
<form class="fect-form ">
  <div class="fect-form-item fect-form-item--top"><label class="fect-form-item__label " for="">
      <!---->
    </label>
    <div class="fect-form-item__content">1</div>
  </div>
  <div class="fect-form-item fect-form-item--left"><label class="fect-form-item__label " for="" style="width: 10px; text-align: left;">
      <!---->
    </label>
    <div class="fect-form-item__content">2</div>
  </div>
  <div class="fect-form-item fect-form-item--right"><label class="fect-form-item__label " for="" style="width: 120px; text-align: right;">
      <!---->
    </label>
    <div class="fect-form-item__content">3</div>
  </div>
</form>
`;
